### PR TITLE
kafka egress

### PR DESCRIPTION
**Pull request by KalmanMeth**

### CODE REVIEW

**Comment by eranra** [comment]
@KalmanMeth I miss the update to docs ... if you run `make build` this should create the API doc file from the changes in the API

**Comment by KalmanMeth** [comment]
@eranra Added section on kafka encode to docs.

File: cmd/flowlogs2metrics/main.go [hunk 1]
```diff
@@ -136,10 +136,11 @@ func initFlags() {
 	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Transform, "pipeline.transform", "[{\"type\": \"none\"}]", "Transforms (list) API")
 	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Extract.Type, "pipeline.extract.type", "", "Extract type: aggregates, none")
 	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Extract.Aggregates, "pipeline.extract.aggregates", "", "Aggregates (see docs)")
-	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Encode.Type, "pipeline.encode.type", "", "Encode type: prom, json, none")
 	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Write.Type, "pipeline.write.type", "", "Write type: stdout, none")
 	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Write.Loki, "pipeline.write.loki", "", "Loki write API")
+	rootCmd.PersistentFlags().StringVar(&config.Opt.PipeLine.Encode.Type, "pipeline.encode.type", "", "Encode type: prom, json, none")
```
Comment (eranra) [inline cmd/flowlogs2metrics/main.go:141]: add kafka as another supported type

File: docs/api.md [hunk 2]
```diff
@@ -17,6 +17,14 @@ Following is the supported API format for prometheus encode:
          prefix: prefix added to each metric name
          expirytime: seconds of no-flow to wait before deleting prometheus data item
 </pre>
+## Kafka encode API
+Following is the supported API format for prometheus encode:
+
+<pre>
+ prom:
```
Comment (eranra) [inline docs/api.md:24]: ??? should be `kafka` not `prom`

File: pkg/api/api.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ const TagEnum = "enum"
 
 type API struct {
 	PromEncode       PromEncode       `yaml:"prom" doc:"## Prometheus encode API\nFollowing is the supported API format for prometheus encode:\n"`
+	KafkaEncode      EncodeKafka      `yaml:"prom" doc:"## Kafka encode API\nFollowing is the supported API format for prometheus encode:\n"`
```
Comment (eranra) [inline pkg/api/api.go:28]: s/prom/kafka
s/prometheus/kafka

File: pkg/pipeline/encode/encode_kafka.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	log.Debugf("entering encodeKafka Encode, #items = %d", len(in))
+	var msg kafkago.Message
+	out := make([]interface{}, 0)
+	for _, entry := range in {
+		var entryByteArray []byte
+		entryByteArray, _ = json.Marshal(entry)
+		msg = kafkago.Message{
+			Value: entryByteArray,
+		}
+		err := r.kafkaWriter.WriteMessages(context.Background(), msg)
```
Comment (eranra) [inline pkg/pipeline/encode/encode_kafka.go:49]: @KalmanMeth maybe some "batch" ing parameters? 

The kafka API is for multiple messages ... maybe it make sense to batch pack all messages and send in one command to kafka ???

File: docs/api.md [hunk 5]
```diff
@@ -17,6 +17,14 @@ Following is the supported API format for prometheus encode:
...
+         addr: address of kafka server
+         topic: kafka topic to write to
```
[the remark this answers is not in the document]
Comment (jpinsonneau) [reply docs/api.md:26]: Yes, at least timeouts and balancer type. You can also add max attempts and max batch size / bytes.

Check [our config](https://github.com/netobserv/goflow2-kube-enricher/blob/afcd39efced182449c48b6742f3c53a08eafcad9/pkg/export/kafka/kafka.go#L32):
```go
	var b kafkago.Balancer
	switch cfg.Writer.Balancer {
	case "roundRobin":
		b = &kafkago.RoundRobin{}
	case "leastBytes":
		b = &kafkago.LeastBytes{}
	case "hash":
		b = &kafkago.Hash{}
	case "crc32":
		b = &kafkago.CRC32Balancer{}
	case "murmur2":
		b = &kafkago.Murmur2Balancer{}
	default:
		return nil, fmt.Errorf("the provided kafka balancer is not valid: %s", cfg.Writer.Balancer)
	}

	return &kafkago.WriterConfig{
		Dialer:       config.NewKafkaDialer(cfg),
		Brokers:      cfg.Writer.Brokers,
		Topic:        cfg.Topic,
		Balancer:     b,
		MaxAttempts:  cfg.Writer.MaxAttempts,
		BatchSize:    cfg.Writer.MaxBatchSize,
		BatchBytes:   cfg.Writer.MaxBatchBytes,
		BatchTimeout: cfg.Writer.BatchTimeout,
		ReadTimeout:  cfg.Writer.ReadTimeout,
		WriteTimeout: cfg.Writer.WriteTimeout,
	}, nil
```

File: cmd/flowlogs2metrics/main.go (hunk 1 above)
Comment (KalmanMeth) [reply cmd/flowlogs2metrics/main.go]: done

File: docs/api.md (hunk 2 above)
Comment (KalmanMeth) [reply docs/api.md]: done

File: docs/api.md [hunk 6]
```diff
@@ -17,6 +17,14 @@ Following is the supported API format for prometheus encode:
          prefix: prefix added to each metric name
          expirytime: seconds of no-flow to wait before deleting prometheus data item
 </pre>
+## Kafka encode API
+Following is the supported API format for prometheus encode:
+
+<pre>
+ prom:
+         addr: address of kafka server
```
[the remark this answers is not in the document]
Comment (KalmanMeth) [reply docs/api.md:25]: done

File: docs/api.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (KalmanMeth) [reply docs/api.md:26]: I added parameters: Balancer, WriteTimeout, ReadTimeout, BatchBytes, BatchSize

File: pkg/api/api.go (hunk 3 above)
Comment (KalmanMeth) [reply pkg/api/api.go]: done

File: pkg/pipeline/encode/encode_kafka.go (hunk 4 above)
Comment (KalmanMeth) [reply pkg/pipeline/encode/encode_kafka.go]: done

**Comment by KalmanMeth** [comment]
Addresses issue #11

**Review by eranra — APPROVED** [review]
/lgtm

**Pull request merged**